src/mito2/src/wal/wal_entry_reader.rs

```diff
@@ -12,13 +12,173 @@
 // See the License for the specific language governing permissions and
 // limitations under the License.
 
+use api::v1::WalEntry;
+use futures::StreamExt;
+use prost::Message;
+use snafu::{ensure, ResultExt};
+use store_api::logstore::entry::Entry;
+use store_api::logstore::provider::Provider;
 use store_api::storage::RegionId;
 
-use crate::error::Result;
-use crate::wal::raw_entry_reader::LogStoreNamespace;
+use crate::error::{CorruptedEntrySnafu, DecodeWalSnafu, Result};
+use crate::wal::raw_entry_reader::RawEntryReader;
 use crate::wal::{EntryId, WalEntryStream};
 
-/// [OneshotWalEntryReader] provides the ability to read and decode entries from the underlying store.
-pub(crate) trait OneshotWalEntryReader: Send + Sync {
-    fn read(self, ctx: LogStoreNamespace, start_id: EntryId) -> Result<WalEntryStream>;
+pub(crate) fn decode_raw_entry(raw_entry: Entry) -> Result<(EntryId, WalEntry)> {
+    let entry_id = raw_entry.entry_id();
+    let region_id = raw_entry.region_id();
+    ensure!(raw_entry.is_complete(), CorruptedEntrySnafu { region_id });
+    // TODO(weny): implement the [Buf] for return value, avoid extra memory allocation.
```

Comment:
We need something like this: https://github.com/apache/opendal/blob/dd70b3c706f1689c2fe9455405cbd1e773c483ab/core/src/types/buffer.rs#L114-L122